trio/io/io.py

```diff
@@ -0,0 +1,97 @@
+from functools import singledispatch, wraps
+import io
+
+import trio
+from trio.io import types
+
+
+__all__ = ['open', 'wrap']
+
+
+class ClosingContextManager:
+    def __init__(self, coro):
+        self._coro = coro
+        self._wrapper = None
+
+    async def __aenter__(self):
+        self._wrapper = await self._coro
+        return self._wrapper
+
+    async def __aexit__(self, typ, value, traceback):
+        await self._wrapper.close()
+
+    def __await__(self):
+        return self._coro.__await__()
+
+
+def closing(func):
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        return ClosingContextManager(func(*args, **kwargs))
+    return wrapper
+
+
+@closing
+async def open(file, mode='r', buffering=-1, encoding=None, errors=None,
+               newline=None, closefd=True, opener=None):
+    """Asynchronous version of :func:`~io.open`.
+
+    Returns:
+        An :term:`asynchronous file object` wrapped in an :term:`asynchronous context manager`.
+
+    Example::
+
+        async with trio.io.open(filename) as f:
+            async for line in f:
+                pass
+
+        assert f.closed
+
+    """
+    _file = wrap(await trio.run_in_worker_thread(io.open, file, mode,
+                                                 buffering, encoding, errors, newline, closefd, opener))
+    return _file
+
+
+@singledispatch
```

Comment:
I'd prefer an explicit `if` chain to the use of `singledispatch`, exactly because allowing random modules to change the behavior of a trio function is a great example of spooky action at a distance.

I get that you want to support `pyfakefs.fake_filesystem.FakeFileWrapper`, but if we're OK with mutating global state like `singledispatch` does then surely we can accomplish this just as easily by doing `io.RawIOBase.register(pyfakefs.fake_filesystem.FakeFileWrapper)` or whatever (and if `pyfakefs` doesn't already do this then that's surely a bug on their end?).

...actually looking more closely, I guess `FakeFileWrapper` actually changes which interface it implements depending on the specific object it's wrapping, which is a bit of a mess but also not one that `@singledispatch` can help with.